### PR TITLE
(PA-1139 PA-1169) Install Hiera 5 config files

### PR DESCRIPTION
**Pull request by jcoconnor**
This supersedes #1076 

Adding an additional commit to this to address preserving the hiera yaml files for windows.

### CODE REVIEW

File: configs/components/puppet.rb [hunk 1]
```diff
@@ -236,4 +237,56 @@
   if platform.is_eos?
     pkg.link "#{settings[:sysconfdir]}", "#{settings[:link_sysconfdir]}"
   end
+
+  pkg.install_file "ext/hiera/hiera.yaml", File.join(settings[:puppet_codedir], 'environments', 'production', 'hiera.yaml')
+  pkg.configfile File.join(settings[:puppet_codedir], 'environments', 'production', 'hiera.yaml')
+  pkg.configfile File.join(configdir, 'hiera.yaml')
+
+  unless platform.is_windows?
```
Comment (branan) [inline configs/components/puppet.rb]: I'm not sure it's necessary to wrap this? The old Hiera pre/post-install actions didn't have this conditional and it seemed to work fine over there.

Comment (jcoconnor) [inline configs/components/puppet.rb]: Would prefer the wrapping as the pre/post actions aren't relevant to windows at all and think its better to make it explicit.

Comment (branan) [inline configs/components/puppet.rb]: Reasonable. I'd actually like to fix Vanagon to throw an error on this sometime in the future. Silently ignoring things makes me sad.

**Comment by jcoconnor** [comment]
Artefacts are collecting at: http://builds.delivery.puppetlabs.net/puppet-agent/b170ded5a106a9d7c82fcf8ad6c205ac211abebe/artifacts/

**Comment by jcoconnor** [comment]
@branan I have had to redo the Windows/MSI installation part as it didn't behave [as per the agreed specifications](https://tickets.puppetlabs.com/browse/PA-1139?focusedCommentId=451542&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-451542) 

I think I've got the conditional testing correct (particularly "_If we detect a hiera.yaml in either $confdir or $environment, don’t install a new file in either location_")

This took a few attempts to get it working, so I resorted to fixing the Wix on the fly on the build machine.

I have also taken the opportunity to tidy/squash the commits so I'm running the whole lot through a Dynamic Vanagon build to make sure I haven't broken anything.

Will update this PR with the build locations once this is done.

I have also added  @Iristyle @McdonaldSeanp as reviewers to ensure my Wixology is non-heretical.

**Comment by jcoconnor** [comment]
Added ```Don't merge``` until commit https://github.com/puppetlabs/puppet-agent/pull/1078/commits/beb476725bd2a864225761f2a4dc3715e0bb9ece is removed.

**Comment by jcoconnor** [comment]
Artefacts are gathering at: http://builds.delivery.puppetlabs.net/puppet-agent/beb476725bd2a864225761f2a4dc3715e0bb9ece/artifacts/

**Comment by jcoconnor** [comment]
Had to queue a rebuild - artefacts will be at: http://builds.delivery.puppetlabs.net/puppet-agent/b9053cb6b0603258a42391989981c63263d1690c/artifacts/

**Comment by jcoconnor** [comment]
I have carried out the following testing on ```Windows-2012r2```

1. Fresh Installation - Hiera 5 files installed in ```environment``` & ```confdir```
2. Upgrade from PA 1.10 - Hiera 5 files are not installed because pre-existing ```hiera.yaml``` in ```confdir```
3. Upgrade from PA 1.4 - Hiera 5 files are not installed because pre-existing ```hiera.yaml``` in ```codedir```

**Comment by jcoconnor** [comment]
I have also done a quick test using ```sles-12```

The upgrade operation also preserves the ```/etc/puppetlabs/puppet/hiera.yam``` file if it has been changed.

The Linux upgrade seems to be a bit cleverer in that it is able to detect if the file has actually been changed from the default, whereas the Windows upgrade will not put in the new files if if finds a version ```hiera.yaml``` there.

**Comment by MosesMendoza** [comment]
It would be nice to have an acceptance test added that validates the install (and ideally upgrade) case - not sure if that's possible especially in the timeframe we have left..

File: configs/components/puppet.rb [hunk 2]
```diff
@@ -236,4 +237,56 @@
...
+    old_hiera = File.join(settings[:puppet_codedir], 'hiera.yaml')
+    new_hiera = File.join(configdir, 'hiera.yaml')
```
Comment (MosesMendoza) [inline configs/components/puppet.rb]: @jcoconnor per @geoffnichols's comment on PA-1139, we should not be overwriting `/etc/puppetlabs/puppet/hiera.yaml` if it already exists. So this needs to also backup `$new_hiera` if it exists on pre-install, and then copy it back on post-install, unconditionally

Comment (branan) [inline configs/components/puppet.rb]: We use `config(noreplace)` on RPM, so the package manager handles that for us. I can double-check Debian. For windows we might need additional logic.

**Comment by branan** [comment]
@MosesMendoza We don't currently have any infrastructure in place for puppet-agent upgrade tests outside of PE. I'm not sure that we have the time to set that up, or that it's worth doing when we only have one test case that would use it.

File: configs/components/puppet.rb (hunk 2 above)
Comment (jcoconnor) [inline configs/components/puppet.rb]: Windows ignores this logic completely and implements it in Wix.
I've designed it to check if either of the ```hiera.yaml``` files are pre-existing - if that is so, the hiera5 files are *not* installed.

We could make backup copies etc., but this is more complex so would prefer to stick with the simpler implementation here.

File: resources/windows/wix/properties.wxs.erb [hunk 3]
```diff
@@ -173,5 +173,26 @@
...
+    <!--
+      Check if there is a pre-existing Hiera 3 format file in either:
+        1. C:\ProgramData\PuppetLabs\code
+        2. C:\ProgramData\PuppetLabs\puppet\etc
+      If either file exists, then we need to block the installation of the new format
+      Hiera 5 default files.
+      This Programdata directory is fixed, so its not influenced by a change in
+      INSTALLDIR, so we can use the util:search property instead of a custom action.
+      These properties are used in appdatafiles.wxs.
+    -->
```
Comment (jcoconnor) [inline resources/windows/wix/properties.wxs.erb]: For reference @branan & @MosesMendoza This is the detection/search for pre-existing files in windows which causes ```PUP_PREEXISTING_CODE_HIERA``` or ```PUP_PREEXISTING_CONF_HIERA``` to be populated.

These are used to drive a conditional for the components (see above) to block them installing if either of these properties exist.

File: resources/windows/wix/appdatafiles.wxs [hunk 4]
```diff
@@ -160,6 +156,32 @@
...
+        Permanent="yes"
+        NeverOverwrite="yes"
+        Guid="998A7670-BF42-4F03-B144-4528EADD0CE3"
+        Directory="PuppetConfDir">
+        <CreateFolder />
+        <File
+          Id="HieraConf"
+          KeyPath="yes"
+          Source="SourceDir\CommonAppDataFolder\Puppetlabs\puppet\etc\hiera.yaml" />
+        <Condition>NOT PUP_PREEXISTING_CODE_HIERA AND NOT PUP_PREEXISTING_CONF_HIERA</Condition>
```
Comment (jcoconnor) [inline resources/windows/wix/appdatafiles.wxs]: This conditional blocks the installation of this component (file) if either of these properties exist.

File: configs/components/puppet.rb (hunk 2 above)
Comment (MosesMendoza) [inline configs/components/puppet.rb]: Per conversation in hipchat, - config(noreplace) will still ovewrite the file if the user hasn't edited it. So we still need logic to backup and then restore any existing `/etc/puppetlabs/puppet/hiera.yaml` file

**Comment by jcoconnor** [comment]
@MosesMendoza @branan Does this change look better ?
Putting it through build and test.

**Comment by jcoconnor** [comment]
New artefacts at: http://builds.delivery.puppetlabs.net/puppet-agent/3cada94a12a4c01e0f7f7a1fc94f7a6420a6b77e/artifacts/

**Comment by jcoconnor** [comment]
Rebuilding again - artefacts will be available at: http://builds.delivery.puppetlabs.net/puppet-agent/76d067c903519ff313c15c11c0358e7328a72979/artifacts/

File: configs/components/puppet.rb [hunk 5]
```diff
@@ -235,4 +236,55 @@
...
+
+    # Post-install script to restore old hiera config files if the have been saved.
+    # and remove any extre hiera configuration files that we laid down
+    postinstall = <<-POSTINST
+# Remove any extra hiera config files we laid down if prev config present
+if [ -e #{rmv_hiera} ]; then
+  rm -f #{cnf_hiera}
+  rm -f #{env_hiera}
+  rm -f #{rmv_hiera}
+fi
```
Comment (MosesMendoza) [inline configs/components/puppet.rb]: With the logic in this script, if the file `/etc/puppetlabs/code/hiera.yaml` exists, but the file `/etc/puppetlabs/puppet/hiera.yaml` does not, we will:
- back up, overwrite, and remove, and restore `/etc/puppetlabs/code/hiera.yaml`
- write and then remove `/etc/puppetlabs/hiera.yaml` 

Leaving the host in a state without any `/etc/puppetlabs/hiera.yaml` (it had none to begin with). I was under the impression we wanted to install the new hiera 5 format yaml to `/etc/puppetlabs/puppet/hiera.yaml` if it does not exist. Is that right?

File: configs/components/puppet.rb [hunk 6]
```diff
@@ -235,4 +236,55 @@
...
+    cnf_hiera = File.join(configdir, 'hiera.yaml')
+    env_hiera = File.join(settings[:puppet_codedir], 'environments', 'production', 'hiera.yaml')
+    rmv_hiera = File.join(configdir, 'remove_hiera5_files.rm')
+
+    # Pre-Install script to save copies of existing hiera configuration files.
+    # This also creates "marker files to indicate that the files pre-existed."
+    preinstall = <<-PREINST
+# Backup the old hiera configs if present, so that we
+# can drop them back in place if the package manager
+# tries to remove it.
```
Comment (MosesMendoza) [inline configs/components/puppet.rb]: Can and should we back up and restore `/etc/puppetlabs/code/environments/production/hiera.yaml` ? Ie for users going from 5.0.0 to 5.0.1 ?

File: configs/components/puppet.rb (hunk 5 above)
Comment (MosesMendoza) [inline configs/components/puppet.rb]: Confirmed in discussion on hipchat that the generic rule here is _if any existing hiera.yaml file exists in any of the known locations, do not install any of the new files_

**Comment by jcoconnor** [comment]
Moses,

Can certainly add this to the list of files in case the package manager
should overwrite it.

File: configs/components/puppet.rb (hunk 6 above)
Comment (geoffnichols) [inline configs/components/puppet.rb]: That sounds right to me. /cc @branan

**Comment by jcoconnor** [comment]
👍 to Branan's last commit.

File: configs/components/puppet.rb [hunk 7]
```diff
@@ -235,4 +236,61 @@
...
+
+# Restore the old hiera, if it existed
+if [ -e #{old_hiera}.pkg-old ]; then
+  mv #{old_hiera}{.pkg-old,}
+fi
+if [ -e #{cnf_hiera}.pkg-old ]; then
+  mv #{cnf_hiera}{.pkg-old,}
+fi
+if [ -e #{env_hiera}.pkg-old ]; then
+  mv #{cnf_hiera}{.pkg-old,}
```
Comment (MosesMendoza) [inline configs/components/puppet.rb]: I thinks this was supposed to be `mv #{env_hiera}{.pkg-old,}` ?

**Pull request merged**